src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -48,27 +45,20 @@ public sealed partial class JsonDocument : IDisposable
         {
             Debug.Assert(!utf8Json.IsEmpty);
 
-            // We never have both rented fields.
-            Debug.Assert(extraRentedArrayPoolBytes == null || extraPooledByteBufferWriter == null);
+            // Both rented values better be null if we're not disposable.
+            Debug.Assert(isDisposable ||
+                (extraRentedArrayPoolBytes == null && extraPooledByteBufferWriter == null));
+
+            // Exactly one of the rented values better be non-null if we're disposable.
+            Debug.Assert(!isDisposable ||
+                ((extraRentedArrayPoolBytes != null || extraPooledByteBufferWriter != null) &&
+                (extraRentedArrayPoolBytes == null || extraPooledByteBufferWriter == null)));
 
             _utf8Json = utf8Json;
             _parsedData = parsedData;
-
-            if (_extraRentedArrayPoolBytes != null)
```

Comment:
This is the bug: should be `extraRentedArrayPoolBytes` instead of `_extraRentedArrayPoolBytes`

Reply:
Seems like this is a good opportunity for a compiler warning? The compiler has knowledge that this field isn't assigned inline with the declaration, so it will always be `default(T)` until it's definitely assigned by a ctor.

Reply:
cc @dotnet/roslyn-compiler

Reply:
Just because the field isn't assigned inline doesn't mean it's not assigned. Consider that the compiler has to contend with the fact that methods, including virtual ones invoked by the base type, can be used to assign fields. That is why the compiler does not generally issue "use before assign" warnings. There are just too many ways for that warning to be falsely emitted. 

Sure `System.Object` is a nice well behaving type in .NET Core but the compiler makes no assumptions about that. It could happily decide to invoke `ToString` in the implicit base ctor here which could assign a bunch of fields. 

More realistically though most code would invoke a property in the ctor. Even invoking a property getter can assign fields hence issuing the warning could be a false positive.

For the compiler to issue a warning it would need to assume rational implementations of `System.Object` and limit itself to cases when no instance methods had been invoked, no properties were invoked or more generally the `this` parameter wasn't used in any other function call. Pretty narrow window for the warning and overall difficult for customers to predict when it will or won't help them.